### ISSUE TITLE
thrust::merge fails for > 2.1B keys

### ISSUE THREAD

**Issue by odanivan** [NVIDIA/thrust#1454]
We are benchmarking the performance of thrust::merge. We do not use any custom allocator and also do not issue the merge on a specific stream. We generate two sorted lists on the host, copy them onto the device, and measure the merge duration.

When trying to merge two sorted lists of equal size and a total number of elements > 2.1 billion keys, we encounter the following error: 

```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  merge: failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
Aborted (core dumped)
```

This bug is very similar to [this one](https://github.com/NVIDIA/thrust/issues/1453) where thrust::sort also fails for data sets with numbers of elements bigger than could be represented by an `int`.

Can anyone help with a hotfix?

**Comment by odanivan** [comment]
Update: When we allocate way to much memory than thrust::merge would need, then we don't get a crash but the data is filled with all zeros. This could mean thrust does not correctly calculate how much memory it needs for merging, right?

**Comment by allisonvacanti** [comment]
This is related to NVIDIA/cub#212. Many of the algorithms in Thrust and CUB are not well behaved when the input size exceeds an int32, and this will take more than a hotfix to address. This is a high priority, but our team is very small and it may be a while before we can get to it.

**Comment by odanivan** [comment]
Thank you for your answer! This information also already helps us a lot.

**Comment by jrhemstad** [comment]
Closing in favor of https://github.com/NVIDIA/thrust/issues/1271
